commons/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java

```diff
@@ -0,0 +1,211 @@
+package org.infinispan.commons.configuration.attributes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.infinispan.commons.CacheException;
+
+/**
+ * Attribute. This class implements a configuration attribute value holder. A configuration attribute is defined by an {@link AttributeDefinition}.
+ * An attribute contains an optional value (or defers to its AttributeDefinition for the default value). It is possible to determine whether
+ * a value has been changed with respect to its initial value. An Attribute remains modifiable until it is protected. At which point it can only
+ * be modified if its AttributeDefinition allows it to be mutable. Additionally it is possible to register listeners when values change so that
+ * code can react to these changes.
+ *
+ * @author Tristan Tarrant
+ * @since 7.2
+ */
+public final class Attribute<T> implements Cloneable {
```

Comment:
Instead of relying on `Cloneable`, could we use a copy constructor/factory instead? See Bloch's Effective Java, Item 11.